src/adapters/philips_hue/discovery.rs

```diff
@@ -0,0 +1,166 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! UPnP and NUPnP discovery for Philips Hue
+//!
+
+pub extern crate url;
+
+use serde_json;
+use std::sync::{ Arc, Mutex };
+use std::thread;
+use super::{ HueAction, http, PhilipsHueAdapter };
+use traits::Controller;
+use transformable_channels::mpsc::*;
+use upnp::{ UpnpListener, UpnpManager, UpnpService };
+
+pub struct Discovery<C> {
+    adapter: PhilipsHueAdapter<C>,
+    upnp_manager: Arc<Mutex<Arc<UpnpManager>>>,
+}
+
+impl<C: Controller> Discovery<C> {
+    pub fn new(adapter: PhilipsHueAdapter<C>) -> Self
+    {
+        let upnp = adapter.controller.get_upnp_manager();
+        let listener = PhilipsHueUpnpListener::new(adapter.clone());
+        upnp.add_listener("PhilipsHueTaxonomy".to_owned(), listener);
+        Discovery {
+            adapter: adapter,
+            upnp_manager: Arc::new(Mutex::new(upnp)),
+        }
+    }
+
+    pub fn do_discovery(&self) {
+        self.do_nupnp_discovery();
+        self.do_upnp_discovery();
+    }
+
+    pub fn do_nupnp_discovery(&self) {
+        let controller = self.adapter.controller.clone();
+        let tx = self.adapter.tx.clone();
+        thread::spawn(move || {
+            let nupnp_enabled = controller.get_config().get_or_set_default(
+                "philips_hue", "nupnp_enabled", "true");
+            if nupnp_enabled == "true" {
+                let nupnp_url = controller.get_config().get_or_set_default(
+                    "philips_hue", "nupnp_url", "http://www.meethue.com/api/nupnp");
+                let nupnp_hubs = nupnp_query(&nupnp_url);
+                for nupnp in nupnp_hubs {
+                    let _ = tx.lock().unwrap().send(HueAction::AddHub(nupnp.id.to_owned(),
+                                nupnp.internalipaddress.to_owned()));
+                }
+            }
+        });
+    }
+
+    pub fn do_upnp_discovery(&self) {
+        let upnp = self.upnp_manager.lock().unwrap();
+        // Still wondering which one of these triggers a bridge response:
+        // upnp.search(Some("urn:schemas-upnp-org:device:basic:1".to_owned())).unwrap();
+        // upnp.search(Some("urn:schemas-upnp-org:device:Basic:1".to_owned())).unwrap();
+        // upnp.search(Some("urn:schemas-upnp-org:device:libhue:idl".to_owned())).unwrap();
+        // upnp.search(Some("upnp:rootdevice".to_owned())).unwrap();
+        // upnp.search(Some("uuid:2f402f80-da50-11e1-9b23-00178825681a".to_owned())).unwrap();
+        upnp.search(None).unwrap();
```

Comment:
Callbacks are triggered even without the search, but then I need to assume that some other adapter's search is also triggering the Hue bridges, but disabling the search here would mean introducing an unwanted dependencies.